sql/bloom.go

```diff
@@ -0,0 +1,240 @@
+// database.
+type DBBloomFilter struct {
+	logger    *zap.Logger
+	name      string
+	mtx       sync.Mutex
+	f         *bloom.BloomFilter
+	sel       expr.Statement
+	idCol     expr.Expr
+	minSize   int
+	fp        float64
+	extraCoef float64
+	added     atomic.Int64
+	loaded    atomic.Int64
+	positive  atomic.Int64
+	negative  atomic.Int64
+	loadOnce  sync.Once
+	eg        errgroup.Group
+	cancel    context.CancelFunc
+	addKeyCh  chan []byte
+}
+
+var _ IDSet = &DBBloomFilter{}
+
+// NewDBBloomFilter creates a new Bloom filter that for a database table.
+// tableName is the name of the table, idColumn is the name of the column that contains
+// the IDs, filter is an optional SQL expression that selects the rows to include in the
+// filter, and falsePositiveRate is the desired false positive rate.
+func NewDBBloomFilter(
+	logger *zap.Logger,
+	name, selectExpr, idCol string,
+	minSize int,
+	extraCoef, falsePositiveRate float64,
+) *DBBloomFilter {
+	return &DBBloomFilter{
+		logger:    logger,
+		name:      name,
+		sel:       expr.MustParseStatement(selectExpr),
+		idCol:     expr.MustParse(idCol),
+		minSize:   minSize,
+		extraCoef: extraCoef,
+		fp:        falsePositiveRate,
+	}
+}
+
+func (bf *DBBloomFilter) Name() string {
+	return bf.name
+}
+
+func (bf *DBBloomFilter) countSQL() string {
+	return expr.SelectBasedOn(bf.sel).Columns(expr.CountStar()).String()
+}
+
+func (bf *DBBloomFilter) loadSQL() string {
+	return bf.sel.String()
+}
+
+func (bf *DBBloomFilter) hasSQL() string {
+	return expr.SelectBasedOn(bf.sel).
+		Where(expr.MaybeAnd(
+			expr.WhereExpr(bf.sel),
+			expr.Op(bf.idCol, expr.EQ, expr.Bind()))).
+		String()
+}
+
+// Add adds the specified key to the Bloom filter.
+func (bf *DBBloomFilter) Add(id []byte) {
+	if bf.addKeyCh != nil {
+		bf.addKeyCh <- id
+	}
+}
+
+// Ready returns true if the Bloom filter is started and has finished loading.
+func (bf *DBBloomFilter) Ready() bool {
+	bf.mtx.Lock()
+	defer bf.mtx.Unlock()
+	return bf.f != nil
+}
+
+func (bf *DBBloomFilter) doLoad(db Executor) error {
+	bf.logger.Info("estimating Bloom filter size", zap.String("name", bf.name))
+	count := 0
+	_, err := db.Exec(bf.countSQL(), nil, func(stmt *Statement) bool {
+		count = stmt.ColumnInt(0)
+		return true
+	})
+	if err != nil {
+		return fmt.Errorf("get count of table %s: %w", bf.name, err)
+	}
+	size := int(math.Ceil(float64(count) * bf.extraCoef))
+	if bf.minSize > 0 && size < bf.minSize {
+		size = bf.minSize
+	}
+	f := bloom.NewWithEstimates(uint(size), bf.fp)
+	bf.logger.Info("loading Bloom filter",
+		zap.String("name", bf.name),
+		zap.Int("count", count),
+		zap.Int("actualSize", size),
+		zap.Int("bytes", f.BitSet().BinaryStorageSize()),
+		zap.Float64("falsePositiveRate", bf.fp))
+	var bs []byte
+	nRows, err := db.Exec(bf.loadSQL(), nil, func(stmt *Statement) bool {
+		l := stmt.ColumnLen(0)
+		if cap(bs) < l {
+			bs = make([]byte, l)
+		} else {
+			bs = bs[:l]
+		}
+		stmt.ColumnBytes(0, bs)
+		f.Add(bs)
```

Comment:
How about using `bytes.Buffer` to avoid manual size work:
```suggestion
	var buf bytes.Buffer
	nRows, err := db.Exec(bf.loadSQL(), nil, func(stmt *Statement) bool {
		buf.Reset()
		buf.ReadFrom(stmt.ColumnReader(0))
		f.Add(buf.Bytes())
```